posts/2022/08-27-speed-up-ci-migrations.md

```diff
@@ -0,0 +1,212 @@
+      |> File.stream!(encoding: :utf8)
+      |> Stream.reject(&String.starts_with?(&1, "--"))
+      |> Stream.reject(&String.starts_with?(&1, "SET"))
+      |> Stream.map(&schema_migrations_create_table_if_not_exists/1)
+      |> Stream.map(&schema_migrations_pk_if_not_exists/1)
+      |> Stream.scan(&remove_multiple_newlines/2)
+      |> Stream.drop(4)
+      |> Stream.into(File.stream!(file_path))
+      |> Stream.run()
+
+      File.rm!(file_path <> ".io")
+    end
+  end
+
+  defp remove_multiple_newlines("\n", "\n"), do: ""
+  defp remove_multiple_newlines("\n", ""), do: ""
+  defp remove_multiple_newlines(line, _), do: line
+
+  defp schema_migrations_create_table_if_not_exists(
+         "CREATE TABLE public.schema_migrations" <> line
+       ) do
+    "CREATE TABLE IF NOT EXISTS public.schema_migrations" <> line
+  end
+
+  defp schema_migrations_create_table_if_not_exists(line), do: line
+
+  defp schema_migrations_pk_if_not_exists("ALTER TABLE ONLY public.schema_migrations" <> _ = line) do
+    drop_constraint_if_exists = """
+    ALTER TABLE ONLY public.schema_migrations
+        DROP CONSTRAINT IF EXISTS schema_migrations_pkey;
+    """
+
+    drop_constraint_if_exists <> "\n" <> line
+  end
+
+  defp schema_migrations_pk_if_not_exists(line), do: line
+end
+```
+
+## Mix task to move migration files
+
+Create in `lib/mix/tasks/move_migrations.ex`
+
+```elixir
+defmodule Mix.Tasks.MoveMigrations do
+  @moduledoc """
+  Mix task for moving mix files to the dumped_migrations folder after their
+  changes have been rolled into the structure.sql dump files. This prevents
+  them from being run again but keeps them as a reference.
+  """
+
+  use Mix.Task
+
+  import Mix.Ecto, only: [ensure_repo: 2, parse_repo: 1]
+  import Mix.EctoSQL, only: [source_repo_priv: 1]
+
+  @source_files "/migrations"
+  @destination "/dumped_migrations"
+
+  @shortdoc "Moves all migrations to the dumped_migrations directory"
+  def run(args) do
+    repos = parse_repo(args)
+
+    Enum.each(repos, fn repo ->
+      ensure_repo(repo, args)
+
+      [repo, migration_repo(repo)]
+      |> Enum.uniq()
+      |> Enum.each(&move_migration_files/1)
+    end)
+  end
+
+  defp migration_repo(repo), do: repo.config()[:migration_repo] || repo
+
+  defp move_migration_files(repo) do
+    path = source_repo_priv(repo)
+
+    File.mkdir_p!(path <> @destination)
+    copied_files = File.cp_r!(path <> @source_files, path <> @destination)
+    Enum.each(copied_files, &IO.puts("Moved file #{&1}"))
+
+    (path <> @source_files)
+    |> File.ls!()
+    |> Enum.reject(&String.contains?(&1, ".gitkeep"))
+    |> Enum.each(&File.rm!(path <> @source_files <> "/" <> &1))
+  end
+end
+```
+
+## Final thoughts
+
+Strategically using database structure dumps can dramatically reduce the time developers need to wait while new databases are created from scratch. Because this occurs frequently both locally and on CI it can be a very useful optimization to save developer toil.
```

Comment:
Database structure dumps can ...  Since databases are created from scratch frequently, both locally and on CI, the strategic use of database structure dumps can be ... 


Last question: save developer toil? maybe save developers some precious time?